Framework/API/src/DetectorSearcher.cpp

```diff
@@ -0,0 +1,251 @@
+ * tracing search strategy
+ *
+ * If no detector is found the first parameter of the returned tuple is false
+ *
+ * @param q :: the Qlab vector to find a detector for
+ * @return tuple with data <detector found, detector index>
+ */
+DetectorSearcher::DetectorSearchResult
+DetectorSearcher::searchUsingInstrumentRayTracing(const V3D &q) {
+  const auto direction = convertQtoDirection(q);
+  m_rayTracer->traceFromSample(direction);
+  const auto det = m_rayTracer->getDetectorResult();
+
+  if (!det)
+    return std::make_tuple(false, 0);
+
+  const auto detIndex = m_detInfo.indexOf(det->getID());
+
+  if (m_detInfo.isMasked(detIndex) || m_detInfo.isMonitor(detIndex))
+    return std::make_tuple(false, 0);
+
+  return std::make_tuple(true, detIndex);
+}
+
+/** Find the index of a detector given a vector in Qlab space using a nearest
+ * neighbours search strategy
+ *
+ * If no detector is found the first parameter of the returned tuple is false
+ *
+ * @param q :: the Qlab vector to find a detector for
+ * @return tuple with data <detector found, detector index>
+ */
+DetectorSearcher::DetectorSearchResult
+DetectorSearcher::searchUsingNearestNeighbours(const V3D &q) {
+  const auto detectorDir = convertQtoDirection(q);
+  // find where this Q vector should intersect with "extended" space
+  const auto neighbours =
+      m_detectorCacheSearch->findNearest(Eigen::Vector3d(q[0], q[1], q[2]), 5);
+  if (neighbours.size() == 0)
+    return std::make_tuple(false, 0);
+
+  const auto result = checkInteceptWithNeighbours(detectorDir, neighbours);
+  const auto hitDetector = std::get<0>(result);
+  const auto index = std::get<1>(result);
+
+  if (hitDetector)
+    return std::make_tuple(true, m_indexMap[index]);
+
+  // Tube Gap Parameter specifically applies to tube instruments
+  if (!hitDetector && m_instrument->hasParameter("tube-gap")) {
+    return handleTubeGap(detectorDir, neighbours);
+  }
+
+  return std::make_tuple(false, 0);
+}
+
+/** Handle the tube-gap parameter in tube based instruments.
+ *
+ * This will check for interceptions with the nearest neighbours by "wiggling"
+ * the predicted detector direction slightly.
+ *
+ * @param detectorDir :: the predicted direction towards a detector
+ * @param neighbours :: the NearestNeighbour results to check interception with
+ * @return a detector search result with whether a detector was hit
+ */
+DetectorSearcher::DetectorSearchResult DetectorSearcher::handleTubeGap(
+    const V3D &detectorDir,
+    const Kernel::NearestNeighbours<3>::NearestNeighbourResults &neighbours) {
+  std::vector<double> gaps = m_instrument->getNumberParameter("tube-gap", true);
+  if (!gaps.empty()) {
+    const auto gap = static_cast<double>(gaps.front());
+    // try adding and subtracting tube-gap in 3 q dimensions to see if you can
+    // find detectors on each side of tube gap
+    for (int i = 0; i < 3; i++) {
+      auto gapDir = V3D(0., 0., 0.);
+      gapDir[i] = gap;
```

Comment:
This code was effectively copied from what `Peak` was already doing during the ray tracing step, except now we only look at the nearest neighbours instead of the whole instrument tree. As far as I understand it, this was a quick hack by @VickieLynch to catch missing reflections falling between tubes.

I'd argue that it is *not* correct for a couple of reasons. 
 1. You're correct in that I think the units are not 100% accurate, it's just the difference between the centre of two (random) pixels minus their width. It does not account for individual values or proper  "curved" interpolation between tubes.
2. This will also potentially catch vectors that fall outside of detectors as well as between tubes.

However, I'd argue that this is still necessary because users do notice when reflections they can clearly see appear to be missing. I think the longer term solution is to come up with and agree on what the proper fix for this case should be. I am very open to suggestions.